File: web/src/core/rakaatStateMachine.ts
import { MovementValidator, TumaninahViolation } from "./movementValidator.js";
import { Posture } from "./posture.js";

export interface Segment {
  posture: Posture;
  start: number;
  end: number;
}

export function segmentDuration(s: Segment): number {
  return s.end - s.start;
}

export interface StateMachineOptions {
  validator?: MovementValidator;
  /** A jalsa held at least this long (after min rakaat) marks the end. */
  finalSitSeconds?: number;
  minRakaatForComplete?: number;
}

/** Counts rakaat from a stream of committed postures + timestamps.
 *
 *  Core insight: **each rakaat contains exactly one ruku**, so rakaat count =
 *  the number of times the body enters the ruku posture. A long final sit
 *  (tashahhud) detects the end. Feed it the smoothed posture once per frame. */
export class RakaatStateMachine {
  rakaatCount = 0;
  isComplete = false;
  segments: Segment[] = [];
  violations: TumaninahViolation[] = [];

  private readonly validator: MovementValidator;
  private readonly finalSitSeconds: number;
  private readonly minRakaatForComplete: number;

  private currentPosture = Posture.Unknown;
  private segmentStart = 0;
  private started = false;

  constructor(opts: StateMachineOptions = {}) {
    this.validator = opts.validator ?? new MovementValidator();
    this.finalSitSeconds = opts.finalSitSeconds ?? 6.0;
    this.minRakaatForComplete = opts.minRakaatForComplete ?? 2;
  }

  /** Feed one frame. `posture` should already be smoothed; `t` is seconds. */
  update(posture: Posture, t: number): void {
    if (this.isComplete) return;

    if (!this.started) {
      this.started = true;
      this.currentPosture = posture;
      this.segmentStart = t;
      if (posture === Posture.Ruku) this.rakaatCount += 1;
      return;
    }

    if (posture !== this.currentPosture) {
      this.closeSegment(t);
      if (posture === Posture.Ruku) this.rakaatCount += 1; // one ruku = one rakaat
      this.currentPosture = posture;
      this.segmentStart = t;
    }

    // Long sustained sitting after enough rakaat ⇒ tashahhud / session end.
    if (
      this.currentPosture === Posture.Jalsa &&
      this.rakaatCount >= this.minRakaatForComplete &&
      t - this.segmentStart >= this.finalSitSeconds
    ) {
      this.closeSegment(t);
      this.isComplete = true;
    }
  }

  /** Force-close the session (user taps stop), flushing the open segment. */
  finish(t: number): void {
    if (!this.started || this.isComplete) return;
    this.closeSegment(t);
    this.isComplete = true;
  }

  private closeSegment(end: number): void {
    const seg: Segment = { posture: this.currentPosture, start: this.segmentStart, end };
    this.segments.push(seg);
    const rakaat = Math.max(1, this.rakaatCount); // pre-first-ruku belongs to rakaat 1
    const v = this.validator.check(seg.posture, segmentDuration(seg), rakaat);
    if (v) this.violations.push(v);
  }

  reset(): void {
    this.rakaatCount = 0;
    this.isComplete = false;
    this.segments = [];
    this.violations = [];
    this.currentPosture = Posture.Unknown;
    this.segmentStart = 0;
    this.started = false;
  }
}

File: Sources/RakaatCore/RakaatStateMachine.swift
import Foundation

/// Counts rakaat from a stream of committed postures + timestamps.
///
/// Core insight (see plan): **each rakaat contains exactly one ruku**, so the
/// rakaat count = the number of times the body enters the ruku posture. Sujud
/// presence and long final sitting (tashahhud) confirm structure and detect the
/// end of the session.
///
/// Feed it the *smoothed* posture (from `PostureSmoother`) once per frame.
public final class RakaatStateMachine {

    /// A contiguous run of one posture.
    public struct Segment: Equatable, Sendable {
        public let posture: Posture
        public let start: Double
        public let end: Double
        public var duration: Double { end - start }
    }

    public private(set) var rakaatCount = 0
    public private(set) var isComplete = false
    public private(set) var segments: [Segment] = []
    public private(set) var violations: [TumaninahViolation] = []

    private let validator: MovementValidator
    /// A jalsa held at least this long (after the min rakaat) marks the end.
    private let finalSitSeconds: Double
    private let minRakaatForComplete: Int

    private var currentPosture: Posture = .unknown
    private var segmentStart: Double = 0
    private var started = false

    public init(
        validator: MovementValidator = MovementValidator(),
        finalSitSeconds: Double = 6.0,
        minRakaatForComplete: Int = 2
    ) {
        self.validator = validator
        self.finalSitSeconds = finalSitSeconds
        self.minRakaatForComplete = minRakaatForComplete
    }

    /// Feed one frame. `posture` should already be smoothed; `t` is seconds.
    public func update(posture: Posture, t: Double) {
        guard !isComplete else { return }

        if !started {
            started = true
            currentPosture = posture
            segmentStart = t
            if posture == .ruku { rakaatCount += 1 }
            return
        }

        if posture != currentPosture {
            closeSegment(end: t)
            // Entering ruku from any other posture = one more rakaat.
            if posture == .ruku { rakaatCount += 1 }
            currentPosture = posture
            segmentStart = t
        }

        // Long sustained sitting after enough rakaat ⇒ tashahhud / session end.
        if currentPosture == .jalsa,
           rakaatCount >= minRakaatForComplete,
           (t - segmentStart) >= finalSitSeconds {
            closeSegment(end: t)
            isComplete = true
        }
    }

    /// Force-close the session (e.g. user taps stop), flushing the open segment.
    public func finish(at t: Double) {
        guard started, !isComplete else { return }
        closeSegment(end: t)
        isComplete = true
    }

    private func closeSegment(end: Double) {
        let seg = Segment(posture: currentPosture, start: segmentStart, end: end)
        segments.append(seg)
        // Pre-first-ruku postures belong to rakaat 1.
        let rakaat = max(1, rakaatCount)
        if let v = validator.check(posture: seg.posture, duration: seg.duration, rakaat: rakaat) {
            violations.append(v)
        }
    }

    public func reset() {
        rakaatCount = 0
        isComplete = false
        segments = []
        violations = []
        currentPosture = .unknown
        segmentStart = 0
        started = false
    }
}
